feat(dspy.Flex) + GEPA extension (#10047)
stanfordnlp/dspy: docs/docs/api/modules/Flex.md

## When to Use Flex

- The best decomposition is **unknown or worth searching** — you have a metric and examples, and you'd rather optimize the program's structure than hand-write it.
- Parts of the task are **deterministic** and shouldn't cost an LM call — arithmetic, parsing, lookups, normalization.
- You want the optimizer to **trade accuracy against cost** — e.g. rewarding programs that answer clear cases in code and reserve the LM for genuinely hard ones.
- You generally don't care much about the concrete implementation of the module.

If the structure is already clear, a hand-written `dspy.Module` (optionally optimized with GEPA or MIPROv2 on its instructions) is simpler. `Flex` earns its keep when the module's forward method itself is the thing you want to learn.

## Basic Usage

```python
import dspy

dspy.configure(lm=dspy.LM("openai/gpt-5"))

# Construct it from a signature, like any module.
solve = dspy.Flex("invoice: str -> total_cents: int")

# Runs using baseline (a single dspy.Predict).
result = solve(invoice="2 widgets @ $3.50, shipping $1.00")
print(result.total_cents)
```

> Out of the box, `solve` is just a `dspy.Predict` over the signature, wrapped in a module. The point of `Flex` is what happens when you optimize it (see [Optimizing with GEPA](#optimizing-with-gepa)): GEPA can replace that baseline with, say, a predictor that only extracts quantities and unit prices, and a line of Python that multiplies and sums them.

Reviewer: Can we mention that it would be an RLM if we had tools up here?

# dspy.Flex

`Flex` is a DSPy module whose implementation is *optimizable code* rather than a fixed prompt. You construct it from a signature, and it starts life as a thin baseline over that signature. What makes it different is what an optimizer is allowed to do with it: instead of only rewriting instructions, `dspy.GEPA` can rewrite the module's entire source — splitting the task into focused predictors, folding deterministic steps into plain Python, and authoring its own helper tools.

> Reach for `Flex` when you don't yet know the right *shape* of a solution — how many LM calls it needs, where code should replace a call, how the work should decompose — and you'd rather have the optimizer discover that structure than hand-write it.

Reviewer: ```suggestion
Reach for `Flex` when you are indifferent to the shape of the solution. You'd rather have the optimizer learn the structure of the function than hand-write it.
```
